### ISSUE TITLE
bug report : ospray compilation fails under Cray compiler wrappers

### ISSUE THREAD

**Issue by demarle** [ospray/ospray#132]
The problem is that ospray/embree fails to compile under cray compiler wrappers, to either intel or gnu compiler suites.

To demonstrate,
1) get on a Cray
2) module load craype (should be on by default)
3) module load PrgEnv-yourchoice
4) export CC=$(which cc)
   export CXX=$(which CC)

The last above makes sure cmake goes through craype instead of bypassing and hitting yourchoice compiler directly.

For OSP 1.1.2, ispc binary
with intel as backend, this is the compile failure
 99%] Linking CXX executable ../../../ospTutorial
icpc: warning #10237: -lcilkrts linked in dynamically, static library not available
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::BVHN(embree::PrimitiveType const&, embree::Scene*)'
../../../libembree.so.2: undefined reference to `embree::Bezier1i::type'
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::clearBarrier(embree::BVHN<4>::NodeRef&)'
../../../libembree.so.2: undefined reference to `embree::SubdivPatch1Cached::type'
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::set(embree::BVHN<4>::NodeRef, embree::BBox<embree::Vec3fa> const&, unsigned long)'
../../../libembree.so.2: undefined reference to `embree::SubdivPatch1Eager::type'
../../../libembree.so.2: undefined reference to `embree::TriangleMvMB<4>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::TriangleM<4>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::layoutLargeNodes(unsigned long)'
../../../libembree.so.2: undefined reference to `embree::Bezier1v::type'
../../../libembree.so.2: undefined reference to `embree::BVHNStatistics<4>::BVHNStatistics(embree::BVHN<4>*)'
../../../libembree.so.2: undefined reference to `embree::TriangleM<8>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::BVHNStatistics<4>::str()'
../../../libembree.so.2: undefined reference to `embree::TrianglePairsMv<4>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::preBuild(std::string const&)'
../../../libembree.so.2: undefined reference to `embree::TriangleMv<4>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::BVHN<4>::postBuild(double)'
../../../libembree.so.2: undefined reference to `embree::TriangleMi<4>::Type::Type()'
../../../libembree.so.2: undefined reference to `embree::Object::type'
builddir/ospray/intel64/CMakeFiles/ospTutorial.dir/build.make:106: recipe for target 'ospTutorial' failed
make[2]: *** [ospTutorial] Error 1

with gnu it is:
[  8%] Building CXX object ospray/embree-v2.7.1/kernels/xeon/CMakeFiles/embree_avx.dir/__/common/scene_subdiv_mesh_avx.cpp.o
/users/demarle/Source/OSPRay/v1.1.2/ospray/ospray/embree-v2.7.1/kernels/common/scene_subdiv_mesh_avx.cpp: In member function 'virtual void embree::SubdivMeshAVX::interpolateN(const void*, const unsigned int*, const float*, const float*, size_t, RTCBufferType, float*, float*, float*, size_t)':
/users/demarle/Source/OSPRay/v1.1.2/ospray/ospray/embree-v2.7.1/kernels/common/scene_subdiv_mesh_avx.cpp:150:30: error: 'loadu' is not a member of 'embree::vint8 {aka embree::vint<8>}'
         if (valid) valid1 &= vint8::loadu(&valid[i]) == vint8(-1);
                              ^~~~~
/users/demarle/Source/OSPRay/v1.1.2/ospray/ospray/embree-v2.7.1/kernels/common/scene_subdiv_mesh_avx.cpp:152:34: error: 'loadu' is not a member of 'embree::vint8 {aka embree::vint<8>}'
         interpolateHelper(valid1,vint8::loadu(&primIDs[i]),vfloat8::loadu(&u[i]),vfloat8::loadu(&v[i]),numUVs,buffer, P ? P+i : nullptr, dPdu ? dPdu+i : nullptr, dPdv ? dPdv+i : nullptr,numFloats);
                                  ^~~~~

**Comment by ingowald** [comment]
We've seen this before, in Stuttgart. IIRC the issue was called by the 
cray compiler incorrectly defining some ISA-specific defines (e.g., 
something like defining AVX2 even if it wasn't enabled, causing embree 
to never generate the SSE-only code bcause it always thinks it was 
compiling for AVX) ... but I forgot the details. (Note this is 
essentialy a compiler bug on cray that causes _embree_ to build 
incorrectly; it's not actually related that much to ospray).

Jeff, Johannes - you remember?

Cheers,

Ingo

**Comment by chuckatkins** [comment]
This isn't using the Cray compiler, it's using the Intel compiler.  The Cray Programming Environment is actually a set of wrappers around an underlying compiler driver that adds lots of additional options to optimize for the specific machine and use libraries loaded by existing module environment.  So, for example, if you load PrgEnv-intel, then the CC, cc, and ftn wrappers actually call icc, icpc, and ifort with additional options.  You can pass the -craype-verbose flag to see the actual Intel compiler command that gets executed by the wrapper.

**Comment by johguenther** [comment]
Embree needs full control over ISA-specific compiler options (because selecting the optimal ISA is done at runtime and not at compile time). Thus if the Cray compiler wrappers add e.g. `-xAVX` or `-march=native` it will result in conflicts and consequently linker errors, see also [Embree issue #115](https://github.com/embree/embree/issues/115).
So, one solution would be to use Embree v2.15 (which contains the appropriate fixes) and in OSPRay v1.1.2 enable `OSPRAY_USE_EXTERNAL_EMBREE` and if necessary let `embree_DIR` point to the install or build folder of Embree. Note that newer versions of OSPRay anyway not not include an own copy Embree anymore.

**Comment by chuckatkins** [comment]
That doesn't fix the issue but you're right in that it is an embree problem so I'll move the discussion over there and open another issue referencing this one.

**Comment by johguenther** [comment]
Actually there was a small adaption needed on OSPRay's side as well (to work with the recent changes in Embree [#118](/embree/embree/issues/118)), which is now available in `devel` 93ec2e53db15980291454153e80db614c8d399c9.

@chuckatkins @demarle can you please confirm it's working now in CPE?

**Comment by chuckatkins** [comment]
Please re-open and leave open until successful CPE functionality can be verified.

**Comment by johguenther** [comment]
Any updates? Does it work now in CPE?

**Comment by Twinklebear** [comment]
I've built and run OSPRay successfully on Theta using the Cray wrappers. To build OSPRay (including the MPI functionality) everything must go through the cray wrappers, so I set:

```
export CC=cc
export CXX=CC
```

to pick the cray wrappers for the C and C++ compilers, and run CMake and explicitly pass the Cray wrappers again as the MPI compilers:

```
cmake -DMPI_C_COMPILER=cc -DMPI_CXX_COMPILER=CC ...
```

Setting the MPI compilers is required for `aprun` to recognize your program as an MPI program. Everything which links or touches MPI must also go through the Cray MPI wrappers, or you'll have some conflicts with the wrong MPI being pulled in.
